Added dedicated register interface

christian-herber-nxp opened the pull request:
- This change is intended to be compatible to the previous definition
    - It adds an option for processors with deeper pipelines to signal the register values in a later pipeline stage than the issue handshake
    - This reduces the register file accesses
    - This can improve performance by allowing register operands to be supplied out of order

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -150,7 +151,7 @@ This non-compressed instruction is then attempted for offload via the issue inte
 The external |coprocessor| can decide to accept or reject the instruction offload. In case of acceptation the |coprocessor|
 will further handle the instruction. In case of rejection the core will raise an illegal instruction exception. 
-As part of the issue interface transaction the core provides the instruction and required register file operand(s) to the |coprocessor|. If
+Subsequently, the core provides the required register file operand(s) to the |coprocessor| via the register interface. If
```
I don't like the word 'Subsequently' here. It should be something like 'Coincidentally or subsequentially'

christian-herber-nxp replied:
what do you not like about this word? subsequentially does not seem to be an English word.

christian-herber-nxp replied:
ok will reword.

christian-herber-nxp replied:
fixed

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -434,84 +440,50 @@ Issue interface
+  |                        |                        | A |coprocessor| must signal ``dualwrite`` as 0 for non-accepted instructions.                                    |
+  +------------------------+------------------------+------------------------------------------------------------------------------------------------------------------+
+  | ``register_read``      | :ref:`registerflags_t  | Will the |coprocessor| perform require specific registers to be read?                                            |
+  |                        | <registerflags>`       | A |coprocessor| may only request an odd register of a pair, if it also requests the even pair                    |
```
the even pair -> the even register of a pair

christian-herber-nxp replied:
will fix

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
 
 * It can handle the instruction (based on decoding ``instr``).
-* The required source registers are marked valid by the offloading core  (``issue_valid`` is 1 and required bit(s) ``rs_valid`` are 1).
+* There are no structural hazards that would prevent execution
```
I would remove this bullet. If not, then it needs to be explained what structural hazards a coprocessor is supposed to check (as most of these checks should actually be performed by the processor, not the coprocessor).

christian-herber-nxp replied:
I would agree that this bullet is somewhat redundant to the first, but it is more specific. What I am trying to state is that the coprocessor would have to check if the necessary execution units are available. Thinking further, as coprocessors may implement custom registers, there may even be data hazards only visible in the coprocessor. Maybe a good topic for discussion

Silabs-ArjanB replied:
I think those are all details that only matter to the coprocessor itself and are confusing to address in the XIF spec. Of course both the processor and coprocessor needs to deal with potential hazards, but we are not going to list them here.

christian-herber-nxp replied:
Implementation hints can absolutely be useful to understand the specification more quickly.

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
+  | **Signal**                | **Type**        | **Direction**   | **Description**                                                                                                              |
+  |                           |                 | (|processor|)   |                                                                                                                              |
+  +---------------------------+-----------------+-----------------+------------------------------------------------------------------------------------------------------------------------------+
+  | ``register_valid``        | logic           | output          | Register request valid. Indicates that |processor| wants to offload an instruction.                                             |
```
Description not accurate (copy / paste from issue_valid)

christian-herber-nxp replied:
fixed

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
+There are two main scenarios, in how the register interface will be used:
+
+1. A register transaction can be started in the same clock cycle as the issue transaction (``issue_req_valid = register_valid`` and ``issue_req.id = register.id``).
+   In this case, the |processor| will speculatively provide all possible source registers via ``register.rs`` when they become available (signalled via the respective ``valid`` signals). 
```
``valid`` -> ``rs_valid``

christian-herber-nxp replied:
will fix

christian-herber-nxp replied:
fixed

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
+1. A register transaction can be started in the same clock cycle as the issue transaction (``issue_req_valid = register_valid`` and ``issue_req.id = register.id``).
+   In this case, the |processor| will speculatively provide all possible source registers via ``register.rs`` when they become available (signalled via the respective ``valid`` signals). 
+   The |coprocessor| will delay accepting the instruction until all necessary registers are provided, and only then assert ``issue_ready``.
+   The ``rs_valid`` bits are not required to be stable during the transaction.
```
Here it is not clear what 'transaction' means because this interface does not have its own ready signal. You could say transaction is from register_valid to issue_ready.

Why not simply add a register_ready signal and add the following restrictions for the two scenarios:

1) issue_valid == register_valid && issue_ready == register_ready
2) issue_ready == 1

christian-herber-nxp replied:
agree

christian-herber-nxp replied:
fixed

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
+
+There are two main scenarios, in how the register interface will be used:
+
+1. A register transaction can be started in the same clock cycle as the issue transaction (``issue_req_valid = register_valid`` and ``issue_req.id = register.id``).
```
issue_req_valid -> issue_valid

christian-herber-nxp replied:
will fix

christian-herber-nxp replied:
fixed

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
+
+2. For a |processor| which splits the issue and register interface into subsequent pipeline stages (e.g. because it has a dedicated read registers (RR) stage), the registers will be provided after the issue transaction completed. 
+   The |processor| initiates the register transaction once all registers are available. 
+   If the |coprocessor| is able to accept multiple issue transactions before receiving the registers, the register transaction can occur in a different order.
```
Can you add a note on the implications for the coprocessor? E.g. it probably needs buffering for all register operands depending on the number of outstanding accepted instructions or something like that.

christian-herber-nxp replied:
the coprocessor will need a buffer the id provided during issue, the mode, and either the instruction or information derived from it. Register operands do not have to be buffered, because they are provided later. Let's discuss

Silabs-ArjanB replied:
If a coprocessor cannot apply back pressure and it needs to deal with out of order register transactions, then it either needs to provide buffering or limit the number of accepted transactions so that out of order register transactions are prevented.

christian-herber-nxp replied:
correct. it can even have a buffer size of 1. i this case, no new instruction will be accepted in issue transaction before the registers are supplied. Will elaborate on that.

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -947,6 +990,7 @@ The following handshake pairs exist on the eXtension interface:
 
 * ``compressed_valid`` with ``compressed_ready``.
 * ``issue_valid`` with ``issue_ready``.
+* ``register_valid`` with implicit always ready signal.
```
No accurate for scenario 1

christian-herber-nxp replied:
i think the best would be to add the ready signal. An note it can be wired to 1 in the second scenario. The same was done for the rs_valid signals, so then this is aligned.

Silabs-ArjanB replied:
It now states 'register_valid with issue_ready' (should use register_ready instead)

christian-herber-nxp replied:
fixed in f9b96bb3563acc2f46d8058afd5b5e46ba20b1c3

Silabs-ArjanB commented on docs/source/x_ext.rst:
```diff
@@ -531,13 +503,83 @@ The |processor| shall cause an illegal instruction fault when attempting to exec
+  | ``ecs_valid``          | logic                    | Validity of the Extension Context Status.                                                                       |
+  +------------------------+--------------------------+-----------------------------------------------------------------------------------------------------------------+
+
+There are two main scenarios, in how the register interface will be used:
```
This will seemingly lead to two incompatible coprocessor types: One that is allowed to put back pressure on the register interface and one that is not allowed to do that? Can we avoid somehow that we end up with incompatible coprocessors? If in total we get 2 flavors of coprocessors (low-end, high-end), then this is maybe acceptable; if we are going to add more flavors then we are causing fragmentation issues.

christian-herber-nxp replied:
it is a topic we need to address. its on the top of my list.

christian-herber-nxp replied:
I suggest we will address this in a separte PR

christian-herber-nxp commented:
@Silabs-ArjanB: Thanks for your rigorous review and input. If you are satisfied with the resolution, please mark the comment as resolved so we can focus on the open ones.

christian-herber-nxp commented:
@jklockars do you have review comments on this PR?

jklockars commented:
> @jklockars do you have review comments on this PR?

The case where register values are provided later seems clear and OK, from what I can tell.

But I think I might be missing something regarding the other case. It _is_ supposed to be compatible, right, so that any core and any coprocessor should be able to talk to eachother no matter what?

Is it supposed to be decided at build time which type of interface will be used? If so, shouldn't that be part of the interface specification itself as some kind of static control bit? Would both types of interfaces need to be supported by the all coprocessors or all cores? Both?

It seems complicated (at best) to allow for different behaviours at run-.time. Such as allowing one instruction to issue (late register values) and then then forcing the next one to halt waiting for values before being accepted (with the previous one receiving its values before the second one).

christian-herber-nxp commented:
> > @jklockars do you have review comments on this PR?
> 
> The case where register values are provided later seems clear and OK, from what I can tell.
> 
> But I think I might be missing something regarding the other case. It _is_ supposed to be compatible, right, so that any core and any coprocessor should be able to talk to eachother no matter what?
> 

A coprocessor could have an elaboration time parameter to be configurable into either mode (maybe it's a good idea to add one). However, there is no way to force an implementer of the interface to support both.

> Is it supposed to be decided at build time which type of interface will be used? If so, shouldn't that be part of the interface specification itself as some kind of static control bit? Would both types of interfaces need to be supported by the all coprocessors or all cores? Both?
> 
> It seems complicated (at best) to allow for different behaviours at run-.time. Such as allowing one instruction to issue (late register values) and then then forcing the next one to halt waiting for values before being accepted (with the previous one receiving its values before the second one).
> 

Clearly this would not be a run-time parameter

christian-herber-nxp commented:
I have added a parameter for the split between issue and register interface. From my point of view, this completes this PR. I have squashed the commits, and the PR should be ready for approval
